File: src/tools/validate-eu-compliance.ts
/**
 * validate_eu_compliance — Check EU alignment status for an Hungarian statute.
 */

import type Database from '@ansvar/mcp-sqlite';
import { euAvailable } from '../capabilities.js';
import { resolveDocumentId } from '../utils/statute-id.js';
import { generateResponseMetadata, type ToolResponse } from '../utils/metadata.js';

export interface ValidateEUComplianceInput {
  document_id: string;
  eu_document_id?: string;
}

interface EUComplianceResult {
  document_id: string;
  document_title: string;
  compliance_status: 'compliant' | 'partial' | 'unclear' | 'not_applicable';
  eu_references_found: number;
  warnings: string[];
  recommendations: string[];
}

export async function validateEUCompliance(
  db: InstanceType<typeof Database>,
  input: ValidateEUComplianceInput,
): Promise<ToolResponse<EUComplianceResult>> {
  const resolvedId = resolveDocumentId(db, input.document_id);
  if (!resolvedId) {
    return {
      results: {
        document_id: input.document_id,
        document_title: 'Unknown',
        compliance_status: 'not_applicable',
        eu_references_found: 0,
        warnings: [`Document not found: "${input.document_id}"`],
        recommendations: [],
      },
      _metadata: generateResponseMetadata(db),
    };
  }

  const doc = db.prepare(
    'SELECT id, title, status FROM legal_documents WHERE id = ?'
  ).get(resolvedId) as { id: string; title: string; status: string };

  const warnings: string[] = [];
  const recommendations: string[] = [];

  if (!euAvailable(db)) {
    return {
      results: {
        document_id: resolvedId,
        document_title: doc.title,
        compliance_status: 'not_applicable',
        eu_references_found: 0,
        warnings: ['EU references not available in this database tier'],
        recommendations: [],
      },
      _metadata: generateResponseMetadata(db),
    };
  }

  let sql = 'SELECT COUNT(*) as count FROM eu_references WHERE document_id = ?';
  const params: string[] = [resolvedId];

  if (input.eu_document_id) {
    sql += ' AND eu_document_id = ?';
    params.push(input.eu_document_id);
  }

  const euRefCount = (db.prepare(sql).get(...params) as { count: number }).count;

  if (euRefCount === 0) {
    return {
      results: {
        document_id: resolvedId,
        document_title: doc.title,
        compliance_status: 'not_applicable',
        eu_references_found: 0,
        warnings: [],
        recommendations: ['No EU cross-references found for this Hungarian statute. Hungary is an EU Member State; EU references indicate transposition obligations.'],
      },
      _metadata: generateResponseMetadata(db),
    };
  }

  if (doc.status === 'repealed') {
    warnings.push('This statute has been repealed.');
    recommendations.push('Check for replacement legislation.');
  }

  // Check implementation status
  const statuses = db.prepare(
    'SELECT implementation_status, COUNT(*) as count FROM eu_references WHERE document_id = ? GROUP BY implementation_status'
  ).all(resolvedId) as { implementation_status: string | null; count: number }[];

  const statusMap = new Map(statuses.map(s => [s.implementation_status, s.count]));
  const completeCount = statusMap.get('complete') ?? 0;
  const partialCount = statusMap.get('partial') ?? 0;
  const unknownCount = statusMap.get('unknown') ?? 0;

  let compliance_status: 'compliant' | 'partial' | 'unclear' | 'not_applicable';
  if (completeCount > 0 && partialCount === 0 && unknownCount === 0) {
    compliance_status = 'compliant';
  } else if (partialCount > 0) {
    compliance_status = 'partial';
    warnings.push(`${partialCount} EU reference(s) have partial alignment status.`);
  } else {
    compliance_status = 'unclear';
    if (unknownCount > 0) {
      recommendations.push(`${unknownCount} EU reference(s) have unknown alignment status. Manual review recommended.`);
    }
  }

  return {
    results: {
      document_id: resolvedId,
      document_title: doc.title,
      compliance_status,
      eu_references_found: euRefCount,
      warnings,
      recommendations,
    },
    _metadata: generateResponseMetadata(db),
  };
}

File: internal/tools/validate-eu-compliance.go
// validate_eu_compliance — Check EU alignment status for an Hungarian statute.
// Port of src/tools/validate-eu-compliance.ts.

package tools

import (
	"database/sql"
	"encoding/json"
	"fmt"

	"github.com/ryzen3100/magyar-jogszabaly-mcp/internal/statute"
	"github.com/ryzen3100/magyar-jogszabaly-mcp/internal/store"
)

type validateEUComplianceArgs struct {
	DocumentID   *string `json:"document_id"`
	EuDocumentID *string `json:"eu_document_id"`
}

// euComplianceResult mirrors the TS EUComplianceResult (singular result
// object). warnings/recommendations are always present as arrays.
type euComplianceResult struct {
	DocumentID        string   `json:"document_id"`
	DocumentTitle     string   `json:"document_title"`
	ComplianceStatus  string   `json:"compliance_status"`
	EuReferencesFound int      `json:"eu_references_found"`
	Warnings          []string `json:"warnings"`
	Recommendations   []string `json:"recommendations"`
}

// ValidateEUCompliance implements the validate_eu_compliance MCP tool. The
// decision ladder reproduces the TS order exactly: unresolved doc → EU probe
// fail → zero references → repealed → status distribution.
func ValidateEUCompliance(db *sql.DB, rawArgs json.RawMessage) (any, ResponseMetadata, error) {
	var args validateEUComplianceArgs
	if len(rawArgs) > 0 {
		if err := json.Unmarshal(rawArgs, &args); err != nil {
			return nil, ResponseMetadata{}, err
		}
	}
	if args.DocumentID == nil {
		return nil, ResponseMetadata{}, fmt.Errorf("missing required argument %q", "document_id")
	}

	resolvedID, err := statute.ResolveDocumentId(db, *args.DocumentID)
	if err != nil {
		return nil, ResponseMetadata{}, err
	}
	if resolvedID == "" {
		// The input is echoed exactly as typed (untrimmed).
		return euComplianceResult{
			DocumentID:        *args.DocumentID,
			DocumentTitle:     "Unknown",
			ComplianceStatus:  "not_applicable",
			EuReferencesFound: 0,
			Warnings:          []string{fmt.Sprintf("Document not found: \"%s\"", *args.DocumentID)},
			Recommendations:   []string{},
		}, GenerateResponseMetadata(db), nil
	}

	var docID, docTitle, docStatus string
	if err := db.QueryRow("SELECT id, title, status FROM legal_documents WHERE id = ?", resolvedID).
		Scan(&docID, &docTitle, &docStatus); err != nil {
		return nil, ResponseMetadata{}, err
	}

	if !store.EUAvailable(db, "eu_references") {
		return euComplianceResult{
			DocumentID:        resolvedID,
			DocumentTitle:     docTitle,
			ComplianceStatus:  "not_applicable",
			EuReferencesFound: 0,
			Warnings:          []string{"EU references not available in this database tier"},
			Recommendations:   []string{},
		}, GenerateResponseMetadata(db), nil
	}

	countSQL := "SELECT COUNT(*) as count FROM eu_references WHERE document_id = ?"
	countParams := []any{resolvedID}
	if args.EuDocumentID != nil && *args.EuDocumentID != "" {
		countSQL += " AND eu_document_id = ?"
		countParams = append(countParams, *args.EuDocumentID)
	}
	var euRefCount int
	if err := db.QueryRow(countSQL, countParams...).Scan(&euRefCount); err != nil {
		return nil, ResponseMetadata{}, err
	}

	if euRefCount == 0 {
		return euComplianceResult{
			DocumentID:        resolvedID,
			DocumentTitle:     docTitle,
			ComplianceStatus:  "not_applicable",
			EuReferencesFound: 0,
			Warnings:          []string{},
			Recommendations: []string{
				"No EU cross-references found for this Hungarian statute. Hungary is an EU Member State; EU references indicate transposition obligations.",
			},
		}, GenerateResponseMetadata(db), nil
	}

	warnings := []string{}
	recommendations := []string{}

	if docStatus == "repealed" {
		warnings = append(warnings, "This statute has been repealed.")
		recommendations = append(recommendations, "Check for replacement legislation.")
	}

	// The status distribution is intentionally NOT filtered by
	// eu_document_id — same quirk as the TS original: the count above is
	// filtered, this GROUP BY is not.
	statusRows, err := db.Query(
		"SELECT implementation_status, COUNT(*) as count FROM eu_references WHERE document_id = ? GROUP BY implementation_status",
		resolvedID)
	if err != nil {
		return nil, ResponseMetadata{}, err
	}
	statusCounts := map[string]int{}
	for statusRows.Next() {
		var (
			status sql.NullString
			n      int
		)
		if err := statusRows.Scan(&status, &n); err != nil {
			statusRows.Close()
			return nil, ResponseMetadata{}, err
		}
		key := ""
		if status.Valid {
			key = status.String
		}
		statusCounts[key] = n
	}
	statusRows.Close()
	if err := statusRows.Err(); err != nil {
		return nil, ResponseMetadata{}, err
	}

	completeCount := statusCounts["complete"]
	partialCount := statusCounts["partial"]
	unknownCount := statusCounts["unknown"]

	var complianceStatus string
	switch {
	case completeCount > 0 && partialCount == 0 && unknownCount == 0:
		complianceStatus = "compliant"
	case partialCount > 0:
		complianceStatus = "partial"
		warnings = append(warnings, fmt.Sprintf("%d EU reference(s) have partial alignment status.", partialCount))
	default:
		complianceStatus = "unclear"
		if unknownCount > 0 {
			recommendations = append(recommendations,
				fmt.Sprintf("%d EU reference(s) have unknown alignment status. Manual review recommended.", unknownCount))
		}
	}

	return euComplianceResult{
		DocumentID:        resolvedID,
		DocumentTitle:     docTitle,
		ComplianceStatus:  complianceStatus,
		EuReferencesFound: euRefCount,
		Warnings:          warnings,
		Recommendations:   recommendations,
	}, GenerateResponseMetadata(db), nil
}
